scale/codec.go

```diff
@@ -70,47 +71,41 @@ func (pe Encoder) PushByte(b byte) error {
 //   zL zL zL 10 / zM zM zM zL / zM zM zM zM / zH zH zH zM					(2**14 ... 2**30 - 1)	(u16, u32)  low LMMH high
 //   nn nn nn 11 [ / zz zz zz zz ]{4 + n}									(2**30 ... 2**536 - 1)	(u32, u64, u128, U256, U512, U520) straight LE-encoded
 // Rust implementation: see impl<'a> Encode for CompactRef<'a, u64>
-func (pe Encoder) EncodeUintCompact(v uint64) error {
-
-	// TODO: handle numbers wider than 64 bits (byte slices?)
-	// Currently, Rust implementation only seems to support u128
-
-	if v < 1<<30 {
-		if v < 1<<6 {
-			err := pe.PushByte(byte(v) << 2)
+func (pe Encoder) EncodeUintCompact(v big.Int) error {
+	if v.Uint64() < 1<<30 {
+		if v.Uint64() < 1<<6 {
+			err := pe.PushByte(byte(v.Uint64()) << 2)
 			if err != nil {
 				return err
 			}
-		} else if v < 1<<14 {
-			err := binary.Write(pe.writer, binary.LittleEndian, uint16(v<<2)+1)
+		} else if v.Uint64() < 1<<14 {
+			err := binary.Write(pe.writer, binary.LittleEndian, uint16(v.Uint64()<<2)+1)
 			if err != nil {
 				return err
 			}
 		} else {
-			err := binary.Write(pe.writer, binary.LittleEndian, uint32(v<<2)+2)
+			err := binary.Write(pe.writer, binary.LittleEndian, uint32(v.Uint64()<<2)+2)
 			if err != nil {
 				return err
 			}
 		}
 		return nil
 	}
 
-	n := byte(0)
-	limit := uint64(1 << 32)
-	for v >= limit && limit > 256 { // when overflows, limit will be < 256
-		n++
-		limit <<= 8
-	}
-	if n > 4 {
-		return errors.New("Assertion error: n>4 needed to compact-encode uint64")
+	numBytes := len(v.Bytes())
+	topSixBits := uint8(numBytes - 4)
+	lengthByte := topSixBits<<2 + 3
+
+	if topSixBits > 63 {
```

Comment:
Line 96 can overflow, so we should modify this check to only use the len, i. e.:

```suggestion
	if numBytes - 4 > 63 {
```